src/pstore/pstore.c

```diff
@@ -0,0 +1,582 @@
+    dirent_ct *dc;
+    LIST_FOREACH(dlcl, dc, *list) {
+        free(dc->contents);
+        LIST_REMOVE(dlcl, (*list), dc);
+        free(dc);
+        break; /* re-start loop */
+    }
+}
+
+static int move_file (dirent_ct *dc, const char *subdir) {
+    _cleanup_free_ char *ifdpath;
+    int remove_file = 0;
+    int r = 0;
+
+    ifdpath = path_join(arg_sourcedir, dc->de.d_name);
+
+    if (arg_storage == PSTORE_STORAGE_EXTERNAL)
+    {
+        /* This code copies the file from the pstore to other storage.
+       The rename() syscall is not utilized as it results in the
+       Invalid cross-device link error.
+       In addition, an optional subdirectory can be specified in
+       forming the final destination path.
+        */
+        _cleanup_free_ char *ofdpath;
+        ofdpath = path_join(arg_archivedir, subdir, dc->de.d_name);
+
+        /* Make sure destination exists */
+        r = mkdir_parents(ofdpath, S_IRWXU|S_IRGRP|S_IXGRP|S_IROTH|S_IXOTH /*0755*/);
+        if (r == 0)
+        {
+#if 1
+            r = copy_file_atomic(ifdpath, ofdpath, 0644, 0, 0, COPY_REPLACE);
+            remove_file = (r == 0);
+#else
+            int ofd;
+
+            ofd = open(ofdpath, O_CREAT|O_TRUNC|O_WRONLY, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH/*0644*/);
+            if (ofd < 0)
+            {
+                log_error_errno(errno, "open(%s): %m\n", ofdpath);
+                r = errno;
+            }
+            else
+            {
+                if (dc->contents != NULL)
+                {
+                    ssize_t wr;
+                    wr = write(ofd, dc->contents, dc->statbuf.st_size);
+                    /* Detect write problem */
+                    if (wr != dc->statbuf.st_size)
+                    {
+                        log_error_errno(errno, "write(%s): %m\n", ofdpath);
+                        r = errno;
+                    }
+                    else
+                        remove_file = 1;
+                }
+            }
+            close(ofd);
+// FIX??? update ofd stat.time to same info as dc.stat.time?
+#endif
+        }
+        else
+        {
+            log_error_errno(errno, "mkdir_parents(): %m\n");
+            r = errno;
+        }
+    }
+    else
+    if (arg_storage == PSTORE_STORAGE_JOURNAL)
+    {
+        //sd_journal_send("MESSAGE=%s", dc->de.d_name, NULL);
+        if (dc->is_binary)
+        {
+            //WITH_BINARY= but how to know end of binary blob if contained in 0 terminated string?
+        }
+        else
+        {
+            r = sd_journal_send("MESSAGE=File %s:\n%s", dc->de.d_name, dc->contents, NULL);
+        }
+        /* NOTE: If journald not running, r always 0 */
+        remove_file = (r == 0);
+    }
+
+    /* If file copied properly, remove it from pstore */
+    if ((r == 0) && remove_file && arg_unlink)
+    {
+        unlink(ifdpath);
+    }
+
+    return r;
+}
+
+static void write_dmesg (const char *dmesg, ssize_t size, const char *id) {
```

Comment:
why use ssize_t here? why not size_t i.e. why signed?

Reply:
changed to size_t